spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/JdbcLockRegistry.java

```diff
@@ -0,0 +1,232 @@
+ * transactions.
+ *
+ * @author Dave Syer
+ *
+ */
+public class JdbcLockRegistry implements ExpirableLockRegistry {
+
+	private final Map<String, JdbcLock> locks = new HashMap<String, JdbcLock>();
+
+	private JdbcClient client;
+
+	public JdbcLockRegistry(JdbcClient client) {
+		this.client = client;
+	}
+
+	@Override
+	public Lock obtain(Object lockKey) {
+		Assert.isInstanceOf(String.class, lockKey);
+		String path = pathFor((String) lockKey);
+		JdbcLock lock = this.locks.get(path);
+		if (lock == null) {
+			synchronized (this.locks) {
+				lock = this.locks.get(path);
+				if (lock == null) {
+					lock = new JdbcLock(this.client, path);
+					this.locks.put(path, lock);
+				}
+			}
+		}
+		return lock;
+	}
+
+	private String pathFor(String lockKey) {
+		MessageDigest digest;
+		try {
+			digest = MessageDigest.getInstance("MD5");
+		}
+		catch (NoSuchAlgorithmException e) {
+			throw new IllegalStateException("MD5 algorithm not available.  Fatal (should be in the JDK).");
+		}
+		try {
+			byte[] bytes = digest.digest(lockKey.getBytes("UTF-8"));
+			return String.format("%032x", new BigInteger(1, bytes));
+		}
+		catch (UnsupportedEncodingException e) {
+			throw new IllegalStateException("UTF-8 encoding not available.  Fatal (should be in the JDK).");
+		}
+	}
+
+	@Override
+	public void expireUnusedOlderThan(long age) {
+		synchronized (this.locks) {
+			Iterator<Entry<String, JdbcLock>> iterator = this.locks.entrySet().iterator();
+			long now = System.currentTimeMillis();
+			while (iterator.hasNext()) {
+				Entry<String, JdbcLock> entry = iterator.next();
+				JdbcLock lock = entry.getValue();
+				if (now - lock.getLastUsed() > age && !lock.isAcquiredInThisProcess()) {
+					iterator.remove();
+				}
+			}
+		}
+	}
+
+	private static final class JdbcLock implements Lock {
+
+		private final JdbcClient mutex;
+
+		private final String path;
+
+		private volatile long lastUsed = System.currentTimeMillis();
+
+		private ReentrantLock delegate = new ReentrantLock();
+
+		private JdbcLock(JdbcClient client, String path) {
+			this.mutex = client;
+			this.path = path;
+		}
+
+		public long getLastUsed() {
+			return this.lastUsed;
+		}
+
+		@Override
+		public void lock() {
+			this.delegate.lock();
+			try {
+				boolean acquired = this.mutex.acquire(this.path);
+				if (acquired) {
+					this.lastUsed = System.currentTimeMillis();
+				}
+			}
+			catch (Exception e) {
+				this.delegate.unlock();
+				throw new RuntimeException("Failed to aquire mutex at " + this.path, e);
+			}
+		}
+
+		@Override
+		public void lockInterruptibly() throws InterruptedException {
+			boolean locked = false;
+			if (this.delegate.isHeldByCurrentThread()) {
+				locked = true;
+			}
+			this.delegate.lockInterruptibly();
+			if (locked) {
+				return;
+			}
+			// this is a bit ugly, but...
+			while (!locked) {
+				try {
+					locked = doLock();
+				}
+				catch (Exception e) {
+					this.delegate.unlock();
+					throw e;
```

Comment:
Thank you, @dsyer , for the contribution!

I'm afraid that this doesn't fit to the current 4.3 GA dead line this Friday.
Just after that we will switch to 5.0 based on Java 8 already, so this issue may go away already :smile: 

```
:spring-integration-jdbc:compileJava/home/travis/build/spring-projects/spring-integration/spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/JdbcLockRegistry.java:158: error: unreported exception Exception; must be caught or declared to be thrown

					throw e;
```

Here is a JIRA to track: https://jira.spring.io/browse/INT-4048

Let us know if that works for you